Python3 error in remote tests

When running remote tests, I get the following error in Python 3

```python
url = 'http://guaix.fis.ucm.es/test.tar.gz'
cache = <numina.tests.diskcache.NuminaDiskCache object at 0x7ff864ca0f28>
bsize = 2048

    def download_cache(url, cache, bsize=BLOCK):
        hh = hashlib.md5()
>       hh.update(url)
E       TypeError: Unicode-objects must be encoded before hashing
```

--- a/numina/core/metaclass.py
+++ b/numina/core/metaclass.py
@@ -144,6 +144,11 @@ class MapStoreType(StoreType):
         # Added due to http://bugs.python.org/issue23572
         return True
 
+    def __nonzero__(self):
+        # Added due to http://bugs.python.org/issue23572
+        return True
+
+
 # Register as a mapping
 collections.Mapping.register(MapStoreType)  # @UndefinedVariable
 

--- a/numina/enum.py
+++ b/numina/enum.py
@@ -71,7 +71,10 @@ class EnumType(type):
         return self.__members__[name]
 
     def __getattr__(self, name):
-        return self.__members__[name]
+        try:
+            return self.__members__[name]
+        except KeyError:
+            raise AttributeError(name)
 
     def __iter__(self):
         return six.itervalues(self.__members__)

--- a/numina/store/default.py
+++ b/numina/store/default.py
@@ -56,7 +56,7 @@ def _(tag, obj, where):
 
 
 @dump.register(DataProductType)
-def _(tag, obj, where):    
+def _(tag, obj, where):
     return obj
 
 
@@ -65,7 +65,7 @@ def _(tag, obj, where):
     return obj
 
 @dump.register(ArrayType)
-def _(tag, obj, where):    
+def _(tag, obj, where):
     return dump.registry[numpy.ndarray](tag, obj, where)
 
 

--- a/numina/tests/diskcache.py
+++ b/numina/tests/diskcache.py
@@ -20,6 +20,7 @@
 import os
 import shutil
 
+import six
 from six.moves import cPickle as pickle
 
 from numina.user.xdgdirs import xdg_cache_home
@@ -29,7 +30,11 @@ class DiskCache(object):
     def __init__(self, cache_dir):
         self._cache = {}
         self.cache_dir = cache_dir
-        self.INDEX_MAP = 'index.map'
+        # Pickle format is different in Py2/Py3
+        if six.PY2:
+            self.INDEX_MAP = 'index.map'
+        else:
+            self.INDEX_MAP = 'index3.map'
         self.index_map = os.path.join(self.cache_dir, self.INDEX_MAP)
 
     def load(self):
@@ -54,7 +59,7 @@ class DiskCache(object):
         # print self._cache
 
     def update_map(self):
-        with open(self.index_map, 'w+') as pk:
+        with open(self.index_map, 'wb+') as pk:
             pickle.dump(self._cache, pk)
 
     def url_is_cached(self, urldigest):

--- a/numina/tests/download.py
+++ b/numina/tests/download.py
@@ -17,11 +17,13 @@
 # along with Numina.  If not, see <http://www.gnu.org/licenses/>.
 #
 
+import hashlib
+from tempfile import NamedTemporaryFile
+
+import six
 from six.moves import urllib_request
 from six.moves import urllib_error
-import hashlib
 
-from tempfile import NamedTemporaryFile
 
 BLOCK = 2048
 
@@ -30,6 +32,7 @@ def download(url, bsize=BLOCK):
     req = urllib_request.Request(url)
     source = urllib_request.urlopen(req)
     #
+    assert False
     with NamedTemporaryFile(delete=False) as fd:
         block = source.read(bsize)
         while block:
@@ -41,7 +44,7 @@ def download(url, bsize=BLOCK):
 
 def download_cache(url, cache, bsize=BLOCK):
     hh = hashlib.md5()
-    hh.update(url)
+    hh.update(six.b(url))
     urldigest = hh.hexdigest()
     update_cache = False
     # print 'urldigest', urldigest
@@ -58,7 +61,7 @@ def download_cache(url, cache, bsize=BLOCK):
     try:
         source = urllib_request.urlopen(req)
         update_cache = True
-        etag = source.headers.dict['etag']
+        etag = source.headers['etag']
     except urllib_error.HTTPError as err:
         if err.code == 304:
             update_cache = False

--- a/numina/tests/download.py
+++ b/numina/tests/download.py
@@ -47,16 +47,16 @@ def download_cache(url, cache, bsize=BLOCK):
     hh.update(six.b(url))
     urldigest = hh.hexdigest()
     update_cache = False
-    print('urldigest', urldigest)
-    print(cache._cache)
+    # print('urldigest', urldigest)
+    # print(cache._cache)
     if cache.url_is_cached(urldigest):
         # Retrieve from cache
         etag = cache.retrieve(urldigest)
-        print('is in cache, etag is', etag)
+        # print('is in cache, etag is', etag)
         req = urllib_request.Request(url)
         req.add_header('If-None-Match', etag)
     else:
-        print('resource not in cache')
+        # print('resource not in cache')
         req = urllib_request.Request(url)
     try:
         source = urllib_request.urlopen(req)
@@ -71,8 +71,6 @@ def download_cache(url, cache, bsize=BLOCK):
 
     #
     with NamedTemporaryFile(delete=False) as fd:
-        print('fd is', fd)
-        print('source is', source)
         block = source.read(bsize)
         while block:
             fd.write(block)

--- a/numina/tests/download.py
+++ b/numina/tests/download.py
@@ -47,16 +47,16 @@ def download_cache(url, cache, bsize=BLOCK):
     hh.update(six.b(url))
     urldigest = hh.hexdigest()
     update_cache = False
-    # print 'urldigest', urldigest
-    # print cache._cache
+    print('urldigest', urldigest)
+    print(cache._cache)
     if cache.url_is_cached(urldigest):
         # Retrieve from cache
         etag = cache.retrieve(urldigest)
-        # print 'is in cache, etag is', etag
+        print('is in cache, etag is', etag)
         req = urllib_request.Request(url)
         req.add_header('If-None-Match', etag)
     else:
-        # print 'resource not in cache'
+        print('resource not in cache')
         req = urllib_request.Request(url)
     try:
         source = urllib_request.urlopen(req)
@@ -65,12 +65,14 @@ def download_cache(url, cache, bsize=BLOCK):
     except urllib_error.HTTPError as err:
         if err.code == 304:
             update_cache = False
-            source = open(cache.cached_filename(urldigest))
+            source = open(cache.cached_filename(urldigest), mode='rb')
         else:
             raise
 
     #
     with NamedTemporaryFile(delete=False) as fd:
+        print('fd is', fd)
+        print('source is', source)
         block = source.read(bsize)
         while block:
             fd.write(block)

